File: 1782-count-pairs-of-nodes/1782-count-pairs-of-nodes.py
class Solution:
    def countPairs(self, n: int, edges: List[List[int]], queries: List[int]) -> List[int]:
        
        nodes = Counter(chain(*edges))
        nodes = {i:nodes[i] if i in nodes else 0 for i in range(1, n+1)}
        edges = Counter(map(lambda x:(min(x), max(x)), edges))
        
        n = 2 + 2*max(nodes.values())
        ans, ctr = [0] * n, Counter(nodes.values())

        for c1, c2 in list(product(ctr, ctr)):
            if c1 > c2: continue
            ans[c1+c2] += ctr[c1]*ctr[c2] if c1!=c2 else ctr[c1]*(ctr[c2]-1)//2

        for i, j in edges:
            sm = nodes[i] + nodes[j]
            ans[sm] -= 1
            ans[sm - edges[(i,j)]]+= 1

        ans = list(accumulate(ans[::-1]))[::-1]

        return [ans[min(q+1, n-1)] for q in queries] 

File: 1782-count-pairs-of-nodes/1782-count-pairs-of-nodes.java
class Solution {
    public int[] countPairs(int n, int[][] edges, int[] qs) {
        int[] cnt = new int[n + 1], sorted_cnt = new int[n + 1], res = new int[qs.length];
        HashMap<Integer, Integer>[] shared = new HashMap[n + 1];
        for (var e : edges) {
            sorted_cnt[e[0]] = cnt[e[0]] = cnt[e[0]] + 1;
            sorted_cnt[e[1]] = cnt[e[1]] = cnt[e[1]] + 1;
            int n1 = Math.min(e[0], e[1]), n2 = Math.max(e[0], e[1]);
            shared[n1] = shared[n1] == null ? new HashMap() : shared[n1];
            shared[n1].put(n2, shared[n1].getOrDefault(n2, 0) + 1);
        }
        Arrays.sort(sorted_cnt);
        for (int q = 0; q < qs.length; ++q) {
            for (int i = 1, j = n; i < j; )
                if (qs[q] < sorted_cnt[i] + sorted_cnt[j])
                    res[q] += (j--) - i;
                else
                    ++i;
            for (int i = 1; i <= n; ++i)
                if (shared[i] != null)
                    for (var p : shared[i].entrySet()) {
                        int j =  p.getKey(), sh_cnt = p.getValue();
                        if (qs[q] < cnt[i] + cnt[j] && qs[q] + sh_cnt >= cnt[i] + cnt[j])
                            --res[q]; 
                    }
        }
        return res;
    }
}